### PR TITLE
*: make partitioned table use statistics

### PR DESCRIPTION
**Pull request by tiancaiamao**
## What have you changed? (mandatory)

Each partition of a partitioned table should use it's own table statistic.
In this PR, `partitionProcessor` will set the `DataSource.tableStatistic`, so the physical plan could use the correct statistic information.

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
manually check the explain result to see the estimated count is right

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Does this PR need to be added to the release notes? (mandatory)
no

## Add a few positive/negative examples (optional)

@lamxTyler @zz-jason

### CODE REVIEW

File: table/tables/partition.go [hunk 1]
```diff
@@ -242,3 +242,7 @@ func (t *partitionedTable) UpdateRecord(ctx sessionctx.Context, h int64, currDat
 	tbl := t.GetPartition(to)
 	return tbl.UpdateRecord(ctx, h, currData, newData, touched)
 }
+
+func (t *partitionedTable) GetID() int64 {
```
Comment (zz-jason) [inline table/tables/partition.go]: irrelevant change?

File: plan/cbo_test.go [hunk 2]
```diff
@@ -451,6 +459,19 @@ func (s *testAnalyzeSuite) TestAnalyze(c *C) {
 			sql:  "analyze table t2 index",
 			best: "Analyze{Index(a),Index(b)}",
 		},
+		// Test partitioned table.
+		{
+			sql:  "select * from t4 where t4.a <= 2",
+			best: "UnionAll{TableReader(Table(t4)->Sel([le(test.t4.a, 2)]))->TableReader(Table(t4)->Sel([le(test.t4.a, 2)]))}",
```
Comment (zz-jason) [inline plan/cbo_test.go]: seems like partition pruning is not working

File: plan/logical_plan_builder.go [hunk 3]
```diff
@@ -1773,12 +1779,16 @@ func (b *planBuilder) buildDataSource(tn *ast.TableName) (LogicalPlan, error) {
 	} else {
 		columns = tbl.Cols()
 	}
+	var statisticTable *statistics.Table
+	if _, ok := tbl.(table.PartitionedTable); !ok {
+		statisticTable = getStatsTable(b.ctx, tbl.Meta(), tbl.Meta().ID)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply plan/logical_plan_builder.go]: If we use `table.Table` as the second parameter:
* tables.Table implements `table.Table`, pass it to `getStatsTable` is ok.
* tables.Partition implements `table.Table`, pass it to `getStatsTable` is ok.
* tables.partitionedTable implements `table.Table`, but pass it to `getStatsTable` will panic.

Then it's not a be general abstraction for this function signature:
```
func getStatsTable(ctx sessionctx.Context, tbl table.Table)
```

File: table/tables/partition.go (hunk 1 above)
Comment (tiancaiamao) [reply table/tables/partition.go]: This change could prevent the potential wrong usage during refactor.

File: plan/cbo_test.go (hunk 2 above)
Comment (tiancaiamao) [reply plan/cbo_test.go]: Partition pruning works right, `t4.a <= 2` will generate the `union all `.
I have thought you mean `Table(t4)->Sel([le(test.t4.a, 2)])` should be `IndexLookup`

Comment (lamxTyler) [reply plan/cbo_test.go]: The plan is correct, it should be table scan.

**Pull request merged**